crates/circuit/src/dag_circuit.rs

```diff
@@ -5958,7 +5856,195 @@ impl DAGCircuit {
+            _ => return Err(DAGCircuitError::new_err("expected node")),
+        };
+
+        let qubit_map = match qubit_map {
+            Some(qubit_map) => qubit_map,
+            None => {
+                let node_qubits = self.qargs_interner.get(node.qubits);
+                let other_qubits = (0..other.num_qubits()).map(Qubit::new);
+                if node_qubits.len() != other_qubits.len() {
+                    return Err(DAGCircuitError::new_err(format!(
+                        "Replacement DAG has {} qubits, expected {}",
+                        other_qubits.len(),
+                        node_qubits.len()
+                    )));
+                }
+                &HashMap::<Qubit, Qubit>::from_iter(other_qubits.zip(node_qubits.iter().copied()))
+            }
+        };
+
+        let clbit_map = match clbit_map {
+            Some(clbit_map) => clbit_map,
+            None => {
+                let node_clbits = self.cargs_interner.get(node.clbits);
+                let other_clbits = (0..other.num_clbits()).map(Clbit::new);
+                if node_clbits.len() != other_clbits.len() {
+                    return Err(DAGCircuitError::new_err(format!(
+                        "Replacement DAG has {} clbits, expected {}",
+                        other_clbits.len(),
+                        node_clbits.len()
+                    )));
+                }
+                &HashMap::<Clbit, Clbit>::from_iter(other_clbits.zip(node_clbits.iter().copied()))
+            }
+        };
+
+        let var_map = match var_map {
+            Some(var_map) => var_map,
+            None => {
+                if self.num_vars() > 0 || other.num_vars() > 0 {
+                    unimplemented!("Inferring variable mapping in substitute_node_with_dag is not implemented yet. Consider using py_substitute_node instead.");
+                    // TODO: implement once additional_wires becomes Python-free
+                }
+                &HashMap::<expr::Var, expr::Var>::new()
+            }
+        };
+
+        let out_map =
+            self.substitute_node_with_graph(node_index, other, qubit_map, clbit_map, var_map)?;
+        self.global_phase = add_global_phase(&self.global_phase, &other.global_phase)?;
+
+        let mut wire_map_dict = HashMap::new();
+        for (source, target) in clbit_map.iter() {
+            let source_bit = other.clbits.get(*source);
+            let target_bit = self.clbits.get(*target);
+            wire_map_dict.insert(source_bit.cloned().unwrap(), target_bit.cloned().unwrap());
+        }
+
+        let variable_mapper = VariableMapper::new(
+            self.cregs.registers().to_vec(),
+            wire_map_dict,
+            var_map.clone(),
+            None,
+        );
+
+        for (old_node_index, new_node_index) in out_map.iter() {
+            let old_node = &other.dag[*old_node_index];
+            if let NodeType::Operation(old_inst) = old_node {
```

Comment:
We can save an indent with something like:

```rust
let NodeType::Operation(old_inst) = old_node else {
    continue;
};
```

It's also nice because it makes it very clear that we're going to ignore anything that doesn't match, whereas with the current style you've got to scroll a ways to see that 🙂.

(I see that you're just moving this code for the most part FWIW lol)